src/customFileParser/main.ts

```diff
@@ -3,77 +3,124 @@ import Sector from './../Sector';
     meshLoader.getGeometry(fileId);
   });
 }
 
-export default async function parseCustomFile(fileBuffer: ArrayBuffer, meshLoader: any, filterOptions?: FilterOptions) {
+export function parseFullCustomFile(
+  fileBuffer: ArrayBuffer,
+  meshLoader: any,
+  filterOptions?: FilterOptions,
+  ): ParseReturn {
   const fileReader = new CustomFileReader(fileBuffer);
-  const idToSectorMap: {[name: string]: Sector} = {};
-  const sceneStats: SceneStats = {
-    numInstancedMeshes: 0,
-    numMergedMeshes: 0,
+  const maps: DataMaps = {
+    treeIndexNodeIdMap: [],
+    colorMap: [],
+    nodeIdTreeIndexMap: new Map(),
+    idToSectorMap: {},
   };
-  const treeIndexNodeIdMap: TreeIndexNodeIdMap = [];
-  const colorMap: ColorMap = [];
+  const compressedData: PerSectorCompressedDataDictionary = {};
 
-  let rootSector = undefined;
-  let uncompressedValues = undefined;
-  const sectorPathToPrimitiveData: {[path: string]: CompressedGeometryData[]} = {};
-  const sectorPathToInstancedMeshData: {[path: string]: CompressedGeometryData} = {};
-  const sectorPathToMergedMeshData: {[path: string]: CompressedGeometryData} = {};
+  // Read root sector
+  const rootSectorLength = fileReader.readUint32();
+  const rootSectorMetadata = fileReader.readSectorMetadata();
+  const rootSector = new Sector(rootSectorMetadata.sectorBBoxMin, rootSectorMetadata.sectorBBoxMax);
+  maps.idToSectorMap[rootSectorMetadata.sectorId] = rootSector;
+  const uncompressedValues = fileReader.readUncompressedValues();
+  compressedData[rootSector.path] = fileReader.readCompressedGeometryData(rootSectorLength);
 
+  // Read remaining sectors
   while (fileReader.location < fileBuffer.byteLength) {
     const sectorStartLocation = fileReader.location;
     const sectorByteLength = fileReader.readUint32();
-    const sectorMetadata = fileReader.readSectorMetadata(sectorByteLength);
+    const sectorMetadata = fileReader.readSectorMetadata();
+    const sector = new Sector(sectorMetadata.sectorBBoxMin, sectorMetadata.sectorBBoxMax);
+    maps.idToSectorMap[sectorMetadata.sectorId] = sector;
+
+    const parentSector = maps.idToSectorMap[sectorMetadata.parentSectorId];
+    if (parentSector !== undefined) {
+      parentSector.addChild(sector);
+      parentSector.object3d.add(sector.object3d);
+    } else { throw Error('Parent sector not found'); }
+    compressedData[sector.path] = fileReader.readCompressedGeometryData(sectorStartLocation + sectorByteLength);
+  }
+
+  return unpackData(rootSector, uncompressedValues, compressedData, maps, filterOptions);
+}
+
+export function parseMultipleCustomFiles(
+  sectorBuffers: ArrayBuffer[],
+  meshLoader: any,
+  filterOptions?: FilterOptions,
+): ParseReturn {
+  const maps: DataMaps = {
+    treeIndexNodeIdMap: [],
+    colorMap: [],
+    nodeIdTreeIndexMap: new Map(),
+    idToSectorMap: {},
+  };
+  const compressedData: PerSectorCompressedDataDictionary = {};
+  let uncompressedValues: undefined | UncompressedValues;
```

Comment:
Very minor thing: for all the cases where you do `undefined | sometype`, can't you write `sometype?` ?

Reply:
That's not valid typescript, unfortunately. At least, 
let blah?: Sector; doesn't work.

Reply:
Ah, you win this time :P